Update the method for determining kernel language

shreve opened the pull request:
The previous algorithm for `getNotebookType()` was based on getting the kernel spec out of the current notebook. This is a problem when format is triggered by, for example, `jupyterlab-autosave-on-focus-change`, immediately after notebook creation. In that case, the language is determined to be `null` and errors arise.

Now the `onSave()` handler waits for the `sessionContext` to be ready, then `getNotebookType()` gets the same kernel spec info from the session rather than the loaded notebook.

krassowski commented on src/formatter.ts:
```diff
@@ -238,6 +238,10 @@ export class JupyterlabNotebookCodeFormatter extends JupyterlabCodeFormatter {
     try {
       this.working = true;
+      if (this.getNotebookType() === null) {
+        this.working = false;
```
Would it be useful to worth some kind of a message here in the console? In general I think that `this.getNotebookType()` result should be presented to the user when the default formatter cannot be found, whether it is because it is `null` or does not match the naming expected.

shreve replied:
The annoying thing about my problem case is that it occurs just as a brand new notebook is opened and an instant later, `getNotebookType()` is correctly populated and no longer null. I experimented with different ways to find this language as the JupyterLab environment knows it before the notebook is created, but they all seemed to have this same problem.

The reason I pursued this PR was because I was getting the "Jupyterlab Code Formatter Error" immediately upon creation, then nothing ended up being broken. I suppose need to do some more work defining replication because I'm not sure if it's just this library, or this library in combination with something else that's triggering it.

My thought was that the other cases where `this.getNotebookType()` are null seem to also suggest being run before the notebook is ready (e.g. `!this.notebookTracker.currentWidget` and `!metadata`).

shreve replied:
I decided to take another approach and have updated the PR, undoing all the changes from the original commit.

shreve commented:
I was pursuing some additional changes around languages for both the notebook and editor formatters, but realized it was a lot to contribute without even asking.

@krassowski @ryantam626 Would you welcome larger code changes? In particular, I added some type definitions, extracted some repeated code, and replaced the hard-coded MIME-type map with the CodeMirror MIME service.

krassowski commented on src/formatter.ts:
```diff
@@ -92,34 +92,14 @@ export class JupyterlabNotebookCodeFormatter extends JupyterlabCodeFormatter {
-      } else if (typeof mode.name === 'string') {
-        return mode.name.toLowerCase();
+  private getNotebookType(): string | null {
+    // Get the language from the current kernel spec
+    const sessionContext = this.notebookTracker.currentWidget?.sessionContext;
+    const kernelName = sessionContext?.session?.kernel?.name;
+    if (kernelName) {
+      const specs = sessionContext.specsManager.specs?.kernelspecs;
+      if (specs && kernelName in specs) {
+        return specs[kernelName]!.language;
```
I think this is a fine approach to try, but I also think we should keep the old code as fallback. This is because there are kernels which will not have an associated session (these are kernels created outside of JupyterLab).

shreve replied:
Oh, that didn't even occur to me 😅 Will do.

krassowski commented on src/index.ts:
```diff
@@ -108,6 +108,7 @@ class JupyterLabCodeFormatter
     state: DocumentRegistry.SaveState
   ) {
     if (state === 'started' && this.config.formatOnSave) {
+      await context.sessionContext.ready;
```
:+1: thanks!

krassowski commented on src/formatter.ts:
```diff
@@ -123,6 +123,16 @@ export class JupyterlabNotebookCodeFormatter extends JupyterlabCodeFormatter {
       }
     }
 
+    // finally, try to get the language from the current session's kernel spec
```
Is it intended that this block will not be reached when metadata are not defined?

krassowski reviewed (approved):
LGTM, thank you @shreve!
